synapse/handlers/sync.py

```diff
@@ -1291,6 +1296,50 @@ async def unread_notifs_for_room_id(
                 sync_config.user.to_string(),
             )
 
+    @cached(max_entries=50000, iterable=True)
+    async def beeper_preview_for_room_id_and_user_id(
+        self, room_id: str, user_id: str, to_key: RoomStreamToken
+    ) -> JsonDict:
+        res = {}
+
+        def _beeper_preview_for_room_id_and_user_id(txn: LoggingTransaction) -> None:
+            sql = """
+            SELECT e.event_id, e.origin_server_ts
+                FROM events AS e \
+                LEFT JOIN redactions as r \
```

Comment:
It's possible to avoid this left join and just use a very specific subquery - might be more expensive than it should be to perform such a join. The stream ordering limit would reduce the complexity significantly.